### PR TITLE
WIP on supporting DI for IHealthCheck instances

### PR DESCRIPTION
**Pull request by bradwilson**
I'm not super happy with how this turned out, so I'm definitely looking for feedback.

TL;DR: We want to be able to support DI for `IHealthCheck` implementors, without introducing DI as a requirement for ASP.NET. A new registration method is added (`HealthCheckBuilder.AddCheck<TCheck>(string name)`) which allows the user to specify a type that should be resolved via DI. The communication between the builder and the service becomes a fairly ugly `Dictionary<string, object>` where the value might be an existing `IHealthCheck` instance, or might be a `Type` that should be resolved via the `IServiceProvider`.

When operating without a DI'd implementation of `IServiceProvider`, we use a simple singleton provider that uses `Activator.CreateInstance` (which means implementations here don't get real DI support). There is, in my mind, an open question about whether ASP.NET should support some kind of user-provided DI, or whether we continue to just directly [create the infrastructure we need](https://github.com/aspnet/HealthChecks/blob/913601fcadbfbd997f32c940d08c08f5f06e00df/src/Microsoft.AspNet.HealthChecks/GlobalHealthChecks.cs#L11-L16) without DI.

/tag @Eilon @glennc @DamianEdwards @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNet.HealthChecks/GlobalHealthChecks.cs [hunk 1]
```diff
@@ -7,11 +8,12 @@ public class GlobalHealthChecks
     {
         static GlobalHealthChecks()
         {
+            // REVIEW: Should we add a way to override the service collection, or just assume no DI here?
```
Comment (davidfowl) [inline src/Microsoft.AspNet.HealthChecks/GlobalHealthChecks.cs:11]: What is this horrific static thing

Comment (bradwilson) [reply src/Microsoft.AspNet.HealthChecks/GlobalHealthChecks.cs:11]: It's called "ASP.NET". Blame ScottGu. ;)
